### DISCUSSION TITLE
Detailed Windows building instructions

### DISCUSSION THREAD

**Discussion by maximka1812** [trufanov-nok/scantailor-universal#125] [General]
It can be nice to have simple detailed instruction on that we need and how to simple build Windows x64 version.

I do not care about Windows XP compatibility or 32 bit versions.

May be also good idea can be to get rid of CMake and other stuff and move to VS project (just for Windows). Can be simple for average non pro users, I mean.

**Comment by trufanov-nok** [comment]
I'm a linux user and STU for me is a linux project. I can't put enouth efforts to test and deploy windows versions. Even launching Windows which would require poweriing off my notebook which could work months without reboot and swapping secondary hdd with some restricted Windows embedded version - is painful.
We can't get rid of cmake as STU requires libpng, libjpeg, libtiff, zlib and maybe something else.. libjpeg2k, libcanberra? can't recall that. And they use cmake. And they must be build with same VS as STU.

If you don't care about WinXP compatibility you may use VS newer than 2013 and Qt newer than 5.6. That's all.
Btw - Qt framework and it's requiements must be built with same VS - that's a cool task too.

I've build these libs with VS 2013 for x86 and x64 and just swapping the folder with binaries of them. Then I'm rebuilding a 2 variandts of STU. And I think Cmake is responsible for launching Nullsoft Installer with .nsi script to produce an installer from all of this at the end.

**Reply by maximka1812** [reply]
I know that you are Linux user, checked previous discussions on forums.

But in reality more than 90% of STU users are Windows users and users that can run x64 app. I do not ask you to change :-), just will try to help here.

Do you always need to build ibpng, libjpeg, libtiff  and similar? You use them as DLLs, at least on Windows. And if we stick to x64 can't we just use prebuild things and some headers?

**Reply by trufanov-nok** [reply]
> Do you always need to build ibpng, libjpeg, libtiff and similar? You use them as DLLs, at least on Windows. And if we stick to x64 can't we just use prebuild things and some headers?

I guess they'll be VCRT dependant: https://www.techpowerup.com/download/visual-c-redistributable-runtime-package-all-in-one/
If you could match all prebuilds together - it would work in theory. I would start from Qt prebuilds.

**Reply by maximka1812** [reply]
It should not make any difference. I mean - we can just use existing prebuild libraries as DLLs that you supply with STU (I mean here release). Same with QT. We can just have header files and necessary DLLs, right?

I aim here only for some simple way so everyone can instantly build ST just by downloading VS (Windows 7+ x64 only!), and not building any libraries for it at all.

**Reply by trufanov-nok** [reply]
> We can just have header files and necessary DLLs, right?

I don't think so. If I recall right you'll need a *.lib files to link with executable. Or it will be a hackery: https://stackoverflow.com/questions/11698228/use-dll-in-c-without-lib

> I aim here only for some simple way so everyone can instantly build ST just by downloading VS (Windows 7+ x64 only!), and not building any libraries for it at all.

That's what would solve DevOps who knows TravisCI.

**Reply by maximka1812** [reply]
Well TravisCI is not simple for any side developer and just another hurdle that you must spend time on.
On constantly updating project that need constant releases it saves time, but is it so in STU case?

I got you considering DLL usage, it is just my histry as Delphi developer, where you can just write unit with headers and it work transparent :-)

**Comment by maximka1812** [comment]
As far as I can see, MS has tools to save you time if you deal with common open source libraries -  vcpkg

https://devblogs.microsoft.com/cppblog/vcpkg-a-tool-to-acquire-and-build-c-open-source-libraries-on-windows/

**Comment by maximka1812** [comment]
So, I am able to compile ScanTailor Advanced (and all required libraries from source, of course). Using MiniGW32-64 and MSYS2.

But ScanTailor Universal still requires more effort :-)
As I understand, Canberra library is not required at all for Windows and should be excluded in CMake if Windows is the target.
So, only openjpeg-2.5.0 must be compiled also (compared to STA). I compiled it from source separately.

Now, can we modify CMake files to make them similar to STA? I mean here STU distributon.
It can be also nice to have same build instructions and BAT files (just need to add single library).

**Reply by trufanov-nok** [reply]
Ok...
You don't need Canberra - it's excluded automatically on Windows during the build.
But you'll need exiv2-0.27 or similar (which is not described in the docs, i'll fix this). I've added this dependancy to keep ICC profile of the original images in STU results.

I've been forced to reboot my notebook, swap the SSD and boot Win to recall how my build folder directory tree looks like. And as I did that anyway I built the most recent STU ver for win64 too. You may download it from this folder; https://drive.google.com/drive/folders/1hT3KUomt1CUGsbXGrBqXYmqr4rAX6wpy?usp=sharing  (the \*18112022\* file). It should have autosave fixed.

As STA deps build process is so easy, I'll fork its build project. I'll add the missing libs. And then will try to make a patch for STU's CMakeList.txt file, so if it's applied it'll be able to be built like STA. I don't wan't to switch on STA build process by default right now. And I really have no time to work on STU right now... But I'll try to make out some working solution during the week or so.

**Reply by maximka1812** [reply]
Thanks.

My current folder structure is made for STA compilation and it is quite nice handy one

libs

> 	Boost
> 	boost_1_80_0
> 	exiv2-0.27.5-MinGW64
> 	jpeg-9e
> 	libpng-1.6.38
> 	openjpeg-2.5.0
> 	Qt
> 	qt-everywhere-opensource-src-5.9.9
> 	scantailor-libs-build
> 	tiff-4.2.0
> 	zlib-1.2.11	

scantailor-advanced
scantailor-universal

exiv2 is from already compiled distribution (made for MiniGW64)
openjpeg in compiled form is available only for MSVC as I see, so this one build from source

By idea with less control we can provide simple few lines batch file and rely on msys2 package manager to download and install all of our libraries, they are available (even minigw64 can be installed via pacman).
But my experience with Microsoft package manager is far from good :-) So I don't use this one also.

**Reply by maximka1812** [reply]
Another difference between STU and STA is translations treatment in Cmake files.

STA uses simpler approach with all related code located in main CMakeLists.txt (plus UpdateTranslations.cmake)

STU here is spread wide - part is in main CMakeLists.txt, part in CMakeLists.txt in src, and part in CMakeLists.txt that is inside src/app

Also STU and STA use different macro names in UpdateTranslations.cmake

**Comment by trufanov-nok** [comment]
Ive spent some time few days ago trying to build STA with msys2 on Win machine. I was able to build all libs, but failed to build Qt 5.7.1 (which is a really last Qt ver that supports WinXP although STA docs says it's Qt 5.6.x) with msys2 and ate some shit with the fact that Qt blocked russian IPs and my old Win system had no Tor or VPN set up. So I decided to postpone that and try again next day with Qt 5.9.9 that you was able to build and that I could download from my linux machine.
And then at some moment I decided to adjust fs partitions on my linux ssd and did a mistake. Now I have a shiny newly installed Kubuntu system on a formatted drive. And this drive lost UEFI partition (or whatever - I'm not a specialist) and I'm 99.9% sure that my hdd with Windows that I usually swap instead of a secondary hdd is not bootable anymore. Thus I'm struggling now with turning back to tests on WIn.
And exactly right now I'n overwhelmed with work: I'v just got a new job, new project and need to spend extra time to put it on rails and push the project team. So, frankly, I don't know when I'll have a time to return to building for Win. I hope on a some weekend.

**Reply by maximka1812** [reply]
Pretty sad story :-)
Only stuff we need is move some parts of ST Advanced Cmake stuff to have ours modified to find and handle libraries located in the ST Advanced folders structure. Plus add two more libraries handling.
Also we need to slowly unify all Cmake files (except small differences that are required, of course) with ST Advanced, I mena besides this modification. This makes it simpler for people.

I am trying to get more handy with CMake. And will try to help.


As for Windows system - I suggest any 128-256GB cheap USB bootable flash drive with proper Windows, on ru-board they have topics about such. It'll work on almost any hardware instantly.

**Comment by noobie-iv** [comment]
@maximka1812 
There is a branch to build for windows:

[https://github.com/noobie-iv/scantailor-universal/tree/build_vs2022x64](https://github.com/noobie-iv/scantailor-universal/tree/build_vs2022x64)

See src\packaging\windows\README.md for instructions.
